File: core/src/main/java/com/netflix/conductor/service/WorkflowBulkService.java
package com.netflix.conductor.service;

import java.util.List;

import org.springframework.validation.annotation.Validated;

import com.netflix.conductor.common.model.BulkResponse;
import com.netflix.conductor.model.WorkflowModel;

import jakarta.validation.constraints.NotEmpty;
import jakarta.validation.constraints.Size;

@Validated
public interface WorkflowBulkService {

    int MAX_REQUEST_ITEMS = 1000;

    BulkResponse<String> pauseWorkflow(
            @NotEmpty(message = "WorkflowIds list cannot be null.")
                    @Size(
                            max = MAX_REQUEST_ITEMS,
                            message =
                                    "Cannot process more than {max} workflows. Please use multiple requests.")
                    List<String> workflowIds);

    BulkResponse<String> resumeWorkflow(
            @NotEmpty(message = "WorkflowIds list cannot be null.")
                    @Size(
                            max = MAX_REQUEST_ITEMS,
                            message =
                                    "Cannot process more than {max} workflows. Please use multiple requests.")
                    List<String> workflowIds);

    BulkResponse<String> restart(
            @NotEmpty(message = "WorkflowIds list cannot be null.")
                    @Size(
                            max = MAX_REQUEST_ITEMS,
                            message =
                                    "Cannot process more than {max} workflows. Please use multiple requests.")
                    List<String> workflowIds,
            boolean useLatestDefinitions);

    BulkResponse<String> retry(
            @NotEmpty(message = "WorkflowIds list cannot be null.")
                    @Size(
                            max = MAX_REQUEST_ITEMS,
                            message =
                                    "Cannot process more than {max} workflows. Please use multiple requests.")
                    List<String> workflowIds);

    BulkResponse<String> terminate(
            @NotEmpty(message = "WorkflowIds list cannot be null.")
                    @Size(
                            max = MAX_REQUEST_ITEMS,
                            message =
                                    "Cannot process more than {max} workflows. Please use multiple requests.")
                    List<String> workflowIds,
            String reason);

    BulkResponse<String> deleteWorkflow(
            @NotEmpty(message = "WorkflowIds list cannot be null.")
                    @Size(
                            max = MAX_REQUEST_ITEMS,
                            message =
                                    "Cannot process more than {max} workflows. Please use multiple requests.")
                    List<String> workflowIds,
            boolean archiveWorkflow);

    BulkResponse<String> terminateRemove(
            @NotEmpty(message = "WorkflowIds list cannot be null.")
                    @Size(
                            max = MAX_REQUEST_ITEMS,
                            message =
                                    "Cannot process more than {max} workflows. Please use multiple requests.")
                    List<String> workflowIds,
            String reason,
            boolean archiveWorkflow);

    BulkResponse<WorkflowModel> searchWorkflow(
            @NotEmpty(message = "WorkflowIds list cannot be null.")
                    @Size(
                            max = MAX_REQUEST_ITEMS,
                            message =
                                    "Cannot process more than {max} workflows. Please use multiple requests.")
                    List<String> workflowIds,
            boolean includeTasks);
}

File: rest/src/services/WorkflowBulkService.ts
import { BulkResponse } from '@conductor/common';
import type { WorkflowService } from './WorkflowService.js';

export class WorkflowBulkService {
  constructor(private readonly workflowService: WorkflowService) {}

  async pauseWorkflow(workflowIds: string[]): Promise<BulkResponse<string>> {
    const response = new BulkResponse<string>();
    for (const wfId of workflowIds) {
      try {
        await this.workflowService.pauseWorkflow(wfId);
        response.appendSuccessResponse(wfId);
      } catch (e) {
        response.appendFailedResponse(wfId, (e as Error).message);
      }
    }
    return response;
  }

  async resumeWorkflow(workflowIds: string[]): Promise<BulkResponse<string>> {
    const response = new BulkResponse<string>();
    for (const wfId of workflowIds) {
      try {
        await this.workflowService.resumeWorkflow(wfId);
        response.appendSuccessResponse(wfId);
      } catch (e) {
        response.appendFailedResponse(wfId, (e as Error).message);
      }
    }
    return response;
  }

  async terminate(workflowIds: string[], reason?: string): Promise<BulkResponse<string>> {
    const response = new BulkResponse<string>();
    for (const wfId of workflowIds) {
      try {
        await this.workflowService.terminateWorkflow(wfId, reason);
        response.appendSuccessResponse(wfId);
      } catch (e) {
        response.appendFailedResponse(wfId, (e as Error).message);
      }
    }
    return response;
  }

  async deleteWorkflow(workflowIds: string[], archiveWorkflow = true): Promise<BulkResponse<string>> {
    const response = new BulkResponse<string>();
    for (const wfId of workflowIds) {
      try {
        await this.workflowService.deleteWorkflow(wfId, archiveWorkflow);
        response.appendSuccessResponse(wfId);
      } catch (e) {
        response.appendFailedResponse(wfId, (e as Error).message);
      }
    }
    return response;
  }

  async restart(workflowIds: string[], useLatestDefinitions = false): Promise<BulkResponse<string>> {
    const response = new BulkResponse<string>();
    for (const wfId of workflowIds) {
      try {
        await this.workflowService.restartWorkflow(wfId, useLatestDefinitions);
        response.appendSuccessResponse(wfId);
      } catch (e) {
        response.appendFailedResponse(wfId, (e as Error).message);
      }
    }
    return response;
  }

  async retry(workflowIds: string[], resumeSubworkflowTasks = false): Promise<BulkResponse<string>> {
    const response = new BulkResponse<string>();
    for (const wfId of workflowIds) {
      try {
        await this.workflowService.retryWorkflow(wfId, resumeSubworkflowTasks);
        response.appendSuccessResponse(wfId);
      } catch (e) {
        response.appendFailedResponse(wfId, (e as Error).message);
      }
    }
    return response;
  }
}
